File: src/Matrix.java
import com.googlecode.lanterna.TextColor;
import com.googlecode.lanterna.screen.TerminalScreen;

import java.util.ArrayList;
import java.util.Random;

public class Matrix {
    private final ArrayList<Line> lines = new ArrayList<>();
    private final TerminalScreen screen;

    public Matrix(TerminalScreen screen) {
        this.screen = screen;
    }

    public void draw(TextColor.RGB color) {
        for (var line : lines) {
            line.draw(screen, color);
        }
    }

    public void advance() {
        for (var line : lines) {
            line.advance();
        }
    }

    public void addLine(Random rng) {
        var line = new Line(rng, (short) screen.getTerminalSize().getColumns());
        lines.add(line);
    }

    public void removeOffScreen() {
        int height = screen.getTerminalSize().getRows();
        lines.removeIf(l -> l.offScreen(height));
    }
}

File: src/matrix.rs
use crate::line::Line;
use colorsys::Hsl;
use crossterm::{
    queue,
    terminal::{Clear, ClearType},
    SynchronizedUpdate,
};
use rand::Rng;
use std::io::{self, Write};

#[derive(Debug, Clone, Default)]
/// All the state that persists between frames.
pub struct Matrix {
    lines: Vec<Line>,
}

impl Matrix {
    pub const fn new() -> Self {
        Matrix { lines: Vec::new() }
    }
    pub fn add_line(&mut self, line: Line) {
        self.lines.push(line)
    }
    pub fn add_random_line(&mut self, rng: &mut impl Rng, width: u16) {
        self.add_line(Line::new(rng, width))
    }
    /// A function that draws a frame, advances all lines and cleans up any lines that go out of
    /// bounds of the screen.
    pub fn update(&mut self, mut writer: impl Write, width: u16, height: u16) -> io::Result<()> {
        let color = Hsl::new(120.0, 100., 100., None);

        // Only update the screen once
        writer.sync_update(|mut writer| -> io::Result<()> {
            queue!(writer, Clear(ClearType::All))?;

            // Only keep the lines that remain in bounds at the end of this opearation
            let mut res = Ok(());
            self.lines.retain_mut(|line| {
                if let Err(err) = line.draw(&mut writer, height, color.clone()) {
                    res = Err(err)
                }
                line.advance();
                line.in_bounds(width, height)
            });
            res
        })?
    }
}
